In [1]:
# HideCell
from IPython.display import HTML

HTML(
    """
    <link href="https://fonts.googleapis.com/css2?family=Roboto:wght@400;700&display=swap" rel="stylesheet">

    <link rel="stylesheet" href="custom.css">
    <script src="custom.js" defer></script>
    <style>
        div.jp-OutputPrompt.jp-OutputArea-prompt {
            display: none !important;
        }
    
        .jp-OutputPrompt {
            display: none !important;
        }

        .jp-Notebook-cell.celltag_remove_cell {
            display: none !important;
        }
    </style>

    """
)

In [2]:
# HideCell
from IPython.display import HTML, Image, display


html_header = """

<header style="position: relative; overflow: hidden; background: #ffffff; font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Inter, Roboto, Arial, sans-serif;">
  <div style="position: absolute; inset: 0; z-index: 0;" aria-hidden="true">
    <svg viewBox="0 0 1400 160" preserveAspectRatio="none" xmlns="http://www.w3.org/2000/svg" style="width:100%; height:100%; display:block;">
      <g stroke="#0c447c" stroke-opacity="0.06" stroke-width="1">
        <line x1="0" y1="40" x2="1400" y2="40"/><line x1="0" y1="80" x2="1400" y2="80"/><line x1="0" y1="120" x2="1400" y2="120"/>
      </g>
      <g fill="#7b828e" fill-opacity="0.28">
        <rect x="120" y="70" width="6" height="64"/><rect x="160" y="55" width="6" height="88"/><rect x="200" y="60" width="6" height="20"/><rect x="240" y="40" width="6" height="55"/><rect x="280" y="65" width="6" height="30"/><rect x="320" y="45" width="6" height="40"/><rect x="360" y="72" width="6" height="22"/><rect x="880" y="58" width="6" height="42"/><rect x="920" y="66" width="6" height="26"/><rect x="960" y="42" width="6" height="50"/><rect x="1000" y="60" width="6" height="32"/><rect x="1040" y="50" width="6" height="44"/><rect x="1080" y="68" width="6" height="24"/><rect x="1120" y="55" width="6" height="38"/>
      </g>
      <g stroke="#7b828e" stroke-opacity="0.28" stroke-width="1">
        <line x1="123" y1="60" x2="123" y2="110"/><line x1="163" y1="48" x2="163" y2="108"/><line x1="203" y1="52" x2="203" y2="86"/><line x1="243" y1="32" x2="243" y2="100"/><line x1="283" y1="58" x2="283" y2="100"/><line x1="323" y1="38" x2="323" y2="90"/><line x1="363" y1="65" x2="363" y2="100"/><line x1="883" y1="50" x2="883" y2="104"/><line x1="923" y1="58" x2="923" y2="96"/><line x1="963" y1="34" x2="963" y2="96"/><line x1="1003" y1="52" x2="1003" y2="96"/><line x1="1043" y1="42" x2="1043" y2="98"/><line x1="1083" y1="60" x2="1083" y2="96"/><line x1="1123" y1="46" x2="1123" y2="97"/>
      </g>
      <polyline fill="none" stroke="#5a6270" stroke-opacity="0.35" stroke-width="1.5"
        points="0,120 60,110 140,95 220,100 300,75 380,85 460,60 540,68 620,45 700,58 780,38 860,50 940,30 1020,42 1100,25 1180,35 1260,18 1400,28"/>
    </svg>
  </div>
  <div style="position: relative; z-index: 1; max-width: 1200px; margin: 0 auto; padding: 18px 24px; display: flex; align-items: center; justify-content: space-between; gap: 24px;">
    <div style="display: flex; align-items: center; gap: 12px;">
      <svg width="30" height="60" viewBox="0 0 30 30" xmlns="http://www.w3.org/2000/svg" role="img" aria-label="Database icon" style="flex-shrink:0;">
        <ellipse cx="15" cy="6" rx="11" ry="4" fill="none" stroke="#7b828e" stroke-width="1.4"/>
        <path d="M4 6 V15 C4 17.2 9 19 15 19 C21 19 26 17.2 26 15 V6" fill="none" stroke="#7b828e" stroke-width="1.4"/>
        <path d="M4 15 V24 C4 26.2 9 28 15 28 C21 28 26 26.2 26 24 V15" fill="none" stroke="#7b828e" stroke-width="1.4"/>
        <ellipse cx="15" cy="15" rx="11" ry="4" fill="none" stroke="#7b828e" stroke-width="1.4"/>
      </svg>
    </div>
  </div>
</header>
"""

display(HTML(html_header))


In [3]:
# HideCell
from IPython.display import HTML, Image, display

# HideCell
display(HTML("""
<script>
document.body.classList.add('preload');
</script>
"""))

In [4]:
# HideCell
HTML("""
<script>
window.addEventListener('load', () => {
    document.body.classList.remove('preload');
});
</script>
""")

In [5]:
# HideCell
from IPython.display import HTML, Image, display
from datetime import date

today_str = date.today().strftime("%B %d, %Y")

html_coverpage_container = f"""
<div id="cover-page-container">
    <div id="title-container">
        <div id="title">Fundamentals Database with SimFin Data, MySQL, and Python</div>
        <div id="meta-data-container">
            <div>{today_str}</div>
            <div>by Aaron Hardy</div>
        </div>
    </div>
    <div id="title-bar"></div>
</div>
"""


display(HTML(html_coverpage_container))

In [ ]:
# HideCell
# GITHUB_PATH = "https://github.com/aaron-san/fundamentals-db-simfin-mysql"
# GITHUB_PATH = "https://github.com/aaron-san/fundamentals-db-simfin-mysql/tree/main/sql/raw"

# Summary

This project involved transforming raw fundamentals data into a structured set of financial ratios that can be used to reveal the underlying health, performance, and risk profile of a company. Raw accounting numbers alone rarely tell a clear story - ratios translate those numbers into standardized, comparable signals. By computing profitability, liquidity, solvency, efficiency, and valuation ratios, the project creates a foundation for deeper analytics, trend detection, and investment‑grade insights. The result is a unified dataset that supports downstream modeling, dashboards, and decision‑making.

# The Data
The raw data was acquired from [SimFin.com](https://www.simfin.com/en/fundamental-data-download/), which requires a free account to download. The free account limits data downloads to dates up to the previous year. 

The raw CSV files were downloaded and each CSV was loaded into a MySQL table. The SQL code used to load each table can be seen in the [Github](https://github.com/aaron-san/fundamentals-db-simfin-mysql/tree/main/sql/raw) repository.

In [6]:
from pathlib import Path
import sys
sys.path.insert(0, "../src")

from db import run_sql_file, run_sql

DATA_DIR = Path("../data")
SQL_DIR = Path("../sql")


In [ ]:
# Load CSV files into SQL database and apply select cleaning

import os
create_raw_table_files = (os.listdir(SQL_DIR / "raw"))

for sql_file in create_raw_table_files:
    run_sql_file(SQL_DIR / "raw" / sql_file)
    # print(f"Created table: {sql_file.split(' - ')[1].split('.sql')[0]}")

The raw data included common industry and sector classifications. The `IndustryId` field maps to the same field in the `us_companies` table.

In [8]:
query = """
WITH cte AS (
SELECT *,
ROW_NUMBER() OVER (PARTITION BY Sector ORDER BY IndustryId) AS rn
FROM industries
)
SELECT Sector, Industry, IndustryId
FROM cte
WHERE rn in (1, 2, 3)
LIMIT 5;
"""

run_sql(query)

,Sector,Industry,IndustryId
0,Basic Materials,Chemicals,110001
1,Basic Materials,Forest Products,110002
2,Basic Materials,Agriculture,110003
3,Business Services,Communication Services,108001
4,Business Services,Consulting,108002


The list of U.S. companies Excel file was inspected in Excel and the file contained rows of data split over several cells in an unpredictable way. In Excel, the CONCAT() function was used to convert all rows to a single concatenated string which was then split again using Data > Text to Columns tool.

The company profile table (`us_companies`) had 37 rows where both the Ticker and Company name were blank (''). 

In [9]:
query = """
SELECT * FROM us_companies;

SELECT COUNT(*)
FROM us_companies
WHERE (
        Ticker = '' OR Ticker IS NULL
    )
    AND (
        CompanyName = '' OR CompanyName IS NULL
    )
    AND (
        CIK = '' OR CIK IS NULL
    );

SELECT "CIK" AS Field, COUNT(*) AS "Missing Values"
FROM us_companies
WHERE CIK = '' OR CIK IS NULL

UNION

SELECT "Ticker" AS Field, COUNT(*) AS "Missing Values"
FROM us_companies
WHERE Ticker = '' OR Ticker IS NULL

UNION

SELECT "CompanyName" AS Field, COUNT(*) AS "Missing Values"
FROM us_companies
WHERE CompanyName = '' OR CompanyName IS NULL

UNION

SELECT "All" AS Field, COUNT(*) AS "Missing Values"
FROM us_companies
WHERE (
        CompanyName = '' OR CompanyName IS NULL
    )
    AND (
        Ticker = '' OR Ticker IS NULL
    )
    AND (
        CIK = '' OR CIK IS NULL
    );

    """
    
summary_missing_values = run_sql(query)
summary_missing_values.style.hide(axis="index")

Field,Missing Values
CIK,2491
Ticker,39
CompanyName,37
All,2


The data includes several financial statement tables. The range of dates for all balance sheet, income statement, and cash flow tables were confirmed below.

In [35]:

query = """
SELECT 
    "us_balance_quarterly" AS table_name,
    MIN(ReportDate) as min_date,
    MAX(ReportDate) as max_date,
    TIMESTAMPDIFF(DAY, MIN(ReportDate), MAX(ReportDate)) / 365.25 AS years
FROM us_balance_quarterly
UNION ALL
SELECT 
    "us_balance_annual" AS table_name,
    MIN(ReportDate) as min_date,
    MAX(ReportDate) as max_date,
    TIMESTAMPDIFF(DAY, MIN(ReportDate), MAX(ReportDate)) / 365.25 AS years
FROM us_balance_annual
UNION ALL
SELECT 
    "us_balance_ttm" AS table_name,
    MIN(ReportDate) as min_date,
    MAX(ReportDate) as max_date,
    TIMESTAMPDIFF(DAY, MIN(ReportDate), MAX(ReportDate)) / 365.25 AS years
FROM us_balance_ttm
UNION ALL
SELECT 
    "us_income_quarterly" AS table_name,
    MIN(ReportDate) as min_date,
    MAX(ReportDate) as max_date,
    TIMESTAMPDIFF(DAY, MIN(ReportDate), MAX(ReportDate)) / 365.25 AS years
FROM us_income_quarterly
UNION ALL
SELECT 
    "us_income_annual" AS table_name,
    MIN(ReportDate) as min_date,
    MAX(ReportDate) as max_date,
    TIMESTAMPDIFF(DAY, MIN(ReportDate), MAX(ReportDate)) / 365.25 AS years
FROM us_income_annual
UNION ALL
SELECT 
    "us_income_ttm" AS table_name,
    MIN(ReportDate) as min_date,
    MAX(ReportDate) as max_date,
    TIMESTAMPDIFF(DAY, MIN(ReportDate), MAX(ReportDate)) / 365.25 AS years
FROM us_income_ttm
UNION ALL
SELECT 
    "us_cashflow_quarterly" AS table_name,
    MIN(ReportDate) as min_date,
    MAX(ReportDate) as max_date,
    TIMESTAMPDIFF(DAY, MIN(ReportDate), MAX(ReportDate)) / 365.25 AS years
FROM us_cashflow_quarterly
UNION ALL
SELECT 
    "us_cashflow_annual" AS table_name,
    MIN(ReportDate) as min_date,
    MAX(ReportDate) as max_date,
    TIMESTAMPDIFF(DAY, MIN(ReportDate), MAX(ReportDate)) / 365.25 AS years
FROM us_cashflow_annual
UNION ALL
SELECT 
    "us_cashflow_ttm" AS table_name,        
    MIN(ReportDate) as min_date,
    MAX(ReportDate) as max_date,
    TIMESTAMPDIFF(DAY, MIN(ReportDate), MAX(ReportDate)) / 365.25 AS years
FROM us_cashflow_ttm;
"""

run_sql(query)

,table_name,min_date,max_date,years
0,us_balance_quarterly,2020-07-31,2025-05-31,4.8323
1,us_balance_annual,2020-07-31,2025-05-31,4.8323
2,us_balance_ttm,2020-07-31,2025-05-31,4.8323
3,us_income_quarterly,2020-07-31,2025-05-31,4.8323
4,us_income_annual,2020-07-31,2025-05-31,4.8323
5,us_income_ttm,2020-07-31,2025-05-31,4.8323
6,us_cashflow_quarterly,2020-07-31,2025-05-31,4.8323
7,us_cashflow_annual,2020-07-31,2025-05-31,4.8323
8,us_cashflow_ttm,2020-07-31,2025-05-31,4.8323


The data spans a period starting *July 31, 2020* and ending *May 31, 2025*, covering about 4.8 years.

## Data Completeness

Before relying on the raw financial statement tables downstream, the completeness of key line items was checked - the percentage of rows where a field is present and non-zero. This flags statement line items that are sparsely populated in the raw SimFin data before ratios are built on top of them. [Code](./sql/)

In [11]:
# Balance sheet field completeness
balance_sheet_validation = run_sql_file(SQL_DIR / "data validation - balance sheet.sql")
balance_sheet_validation.style.hide(axis="index")

TotalRows,CashAndEquiv,AccountsNotesReceivable,Inventories,TotalCurrentAssets,PropertyPlantEquipmentNet,LongTermInvestmentsReceivables,OtherLongTermAssets,TotalAssets,PayablesAccruals,ShortTermDebt,TotalCurrentLiabilities,LongTermDebt,TotalLiabilities,ShareCapitalAdditionalPaidIn,TotalEquity,TotalLiabilitiesEquity,SharesBasic,SharesDiluted
16898,99.5%,78.3%,54.3%,99.9%,93.1%,23.2%,96.6%,100.0%,99.6%,53.0%,99.9%,70.6%,99.9%,98.9%,100.0%,100.0%,98.8%,98.3%


In [12]:
# Income statement field completeness
income_statement_validation = run_sql_file(SQL_DIR / "data validation - income statement.sql")
income_statement_validation.style.hide(axis="index")

TotalRows,Revenue,GrossProfit,OperatingIncomeLoss,NetIncomeCommon
16897,89.2%,77.8%,100.0%,99.9%


In [13]:
# Cash flow statement field completeness
cash_flow_validation = run_sql_file(SQL_DIR / "data validation - cash flows.sql")
cash_flow_validation.style.hide(axis="index")

TotalRows,NetCashFromOperatingActivities,NetIncomeStartingLine,SharesBasic,SharesDiluted
16895,99.9%,99.2%,98.8%,98.3%


Next, the minimum and maximum prices for both `Close` and `AdjClose` were observed. 

In [14]:
query = """
SELECT 
    "Close" AS "Field", ROUND(MIN(Close), 2) AS "Min",
    ROUND(MAX(Close), 2) AS "Max"
FROM us_shareprices_daily
UNION ALL
SELECT 
    "AdjClose" AS "Field", ROUND(MIN(AdjClose), 2) AS "Min",
    ROUND(MAX(AdjClose), 2) AS "Max"
FROM us_shareprices_daily;
"""
price_stats = run_sql(query)
price_stats.style.hide(axis="index")


Field,Min,Max
Close,0.00,99999999.99
AdjClose,0.00,99999999.99


The `Close` price ranged from \\$0 to \\$99,999,999.99. 
The `Adj. Close` price ranged from \\$0 to \\$9,984,240.00.

Tickers where a `Close` or `AdjClose` price for any day exceeding \\$100,000 were removed, since the common investor is not likely to have an investing budget exceeding \\$100,000 for any single ticker. The user can adjust this to fit their budget or use the original `us_shareprices_daily` table.

In [15]:
# Hide
query = """
DROP TABLE IF EXISTS us_shareprices_daily_filtered;

CREATE TABLE us_shareprices_daily_filtered AS

WITH
    filtered_data AS (
        SELECT DISTINCT
            Ticker
        FROM us_shareprices_daily
        WHERE
            Close > 100000
            OR AdjClose > 100000
    )
SELECT *
FROM us_shareprices_daily
WHERE
    Ticker NOT IN (SELECT Ticker FROM filtered_data)
"""

run_sql(query)


In [16]:
prices_daily_filtered = run_sql("SELECT * FROM us_shareprices_daily_filtered LIMIT 5")
prices_daily_filtered.style.hide(axis="index")

Ticker,SimFinId,Date,Open,High,Low,Close,AdjClose,Volume,Dividend,SharesOutstanding
,18728220,2022-07-12,76.1000,134.0000,75.1000,120.5000,120.5000,148681,0.000000,0
,18728220,2022-07-13,97.7000,120.0000,88.0000,88.0000,88.0000,26444,0.000000,0
,18728220,2022-07-14,80.1000,82.6000,63.2100,73.5000,73.5000,13413,0.000000,0
,18728220,2022-07-15,76.4000,78.2000,68.0000,68.0000,68.0000,4699,0.000000,0
,18728220,2022-07-18,64.5000,84.9000,63.0000,67.0000,67.0000,6941,0.000000,0


Next, the daily stock prices were converted to monthly prices. [Code](./sql/derived/)

In [17]:
run_sql_file(SQL_DIR / "derived" / "01 - create table - us_shareprices_monthly.sql")
us_shareprices_monthly = run_sql("SELECT * FROM us_shareprices_monthly")
display(us_shareprices_monthly.tail(5))

,Ticker,Date,Close,AdjClose
300528,ZYNE,2023-06-30,0.3300,0.3300
300529,ZYNE,2023-07-31,0.3600,0.3600
300530,ZYNE,2023-08-31,1.3300,1.3300
300531,ZYNE,2023-09-29,1.2700,1.2700
300532,ZYNE,2023-10-11,1.3000,1.3000


In [18]:
run_sql_file(SQL_DIR / "derived" / "02 - create table - us_shareprices_monthly_returns.sql")
us_shareprices_monthly_returns = run_sql("SELECT * FROM us_shareprices_monthly_returns")
display(us_shareprices_monthly_returns.tail(5))

,Ticker,Date,AdjClose,Close,PrevAdjClose,MonthlyReturn
300528,ZYNE,2023-06-30,0.3300,0.3300,0.3300,0.000000
300529,ZYNE,2023-07-31,0.3600,0.3600,0.3300,0.090909
300530,ZYNE,2023-08-31,1.3300,1.3300,0.3600,2.694444
300531,ZYNE,2023-09-29,1.2700,1.2700,1.3300,-0.045113
300532,ZYNE,2023-10-11,1.3000,1.3000,1.2700,0.023622


## Price Statistics

Finally, a set of price-based statistics was derived from the monthly price series and stored in `us_shareprices_stats`: running maximum, drawdown from that peak, 3/6/12-month moving averages of the adjusted close, and trailing 3-year annualized volatility (computed from monthly log returns). These give a quick read on a ticker's trend and risk profile, independent of the fundamentals-based ratios computed below. [Code](./sql/derived/)

In [19]:
run_sql_file(SQL_DIR / "derived" / "create table - us_shareprices_stats.sql")
us_shareprices_stats = run_sql("SELECT * FROM us_shareprices_stats")
display(us_shareprices_stats.tail(5))

,Ticker,Date,Close,AdjClose,MonthlyReturn,RunningMax,Drawdown,MA_3M,MA_6M,MA_1Y,Annualized_3Yr_Volatility
300528,ZYNE,2023-06-30,0.3300,0.3300,0.0000,5.6800,-0.9420,0.3470,0.4330,0.6200,0.6402
300529,ZYNE,2023-07-31,0.3600,0.3600,0.0870,5.6800,-0.9370,0.3400,0.3820,0.5550,0.6375
300530,ZYNE,2023-08-31,1.3300,1.3300,1.3068,5.6800,-0.7660,0.6730,0.5270,0.5680,1.0082
300531,ZYNE,2023-09-29,1.2700,1.2700,-0.0462,5.6800,-0.7760,0.9870,0.6670,0.6130,1.0064
300532,ZYNE,2023-10-11,1.3000,1.3000,0.0233,5.6800,-0.7710,1.3000,0.8200,0.6610,1.0062


# Ratios & Metrics

## Dates Lookup Table

Dates are an important part of any analysis that uses fundamentals data. Market price dates don't always align with fundamentals dates. Moreover, there are many kinds of dates to consider. Any serious analysis would try to avoid lookback bias where a decision based on information not yet known. To avoid this bias, the stock price date following the fundamentals publish date was collected. 

For backtesting decisions, it is often easier to use a regular month-end price date where the month-end date is the end of the month following the publish date. The month-end date following the publish date was calculated as the last day of the publish-date month or, if the publish date is within 3 days of the month end, the last day of the following month. 


In [20]:
run_sql_file(SQL_DIR / "derived" / "create table - dates_lookup.sql")
dates_lookup = run_sql("SELECT * FROM dates_lookup")
display(dates_lookup.tail(5))

,Ticker,ReportDate,PrevCloseDate,PublishDate,NextMarketDateFollowingPublish,NextMonthEndDateFollowingPublish
47640,ZYXI,2024-03-31,None,2024-04-30,None,None
47641,ZYXI,2024-06-30,None,2024-07-25,None,None
47642,ZYXI,2024-09-30,None,2024-10-24,None,None
47643,ZYXI,2024-12-31,None,2025-03-11,None,None
47644,ZYXI,2025-03-31,None,2025-04-29,None,None


## Miscellaneous Metrics
Next, miscellaneous metrics such as effective tax rate, market cap, total debt, and free cash flow were computed and inserted into a new table, `metrics_misc`

In [21]:
run_sql_file(SQL_DIR / "derived" / "create table - metrics_misc.sql")
metrics_misc = run_sql("SELECT * FROM metrics_misc")
metrics_misc.iloc[120:126]

,Ticker,ReportDate,CashAndEquiv,EffectiveTaxRate,PrevCloseDate,MarketCap,TotalDebt,FreeCashFlow,EnterpriseValue,PayoutRatio,FilledPayoutRatio,TtmAvgPayoutRatio,TickerAveragePayout
120,AAOI,2023-03-31,16103000.0,None,2023-03-31,63806803.9700,130631000.0,-10013000.0,178334803.9700,None,None,None,None
121,AAOI,2023-06-30,21578000.0,None,2023-06-30,175751823.5600,117192000.0,14236000.0,271365823.5600,None,None,None,None
122,AAOI,2023-09-30,21940000.0,None,2023-09-29,359532403.5600,111447000.0,-16912000.0,449039403.5600,None,None,None,None
123,AAOI,2023-12-31,45366000.0,None,2023-12-29,707932520.4000,115198000.0,-4772000.0,777764520.4000,None,None,None,None
124,AAOI,2024-03-31,8433000.0,None,2024-03-28,531698955.4800,111172000.0,-34363000.0,634437955.4800,None,None,None,None
125,AAOI,2024-06-30,10653000.0,None,2024-06-28,326333528.8000,104244000.0,-5076000.0,419924528.8000,None,None,None,None


## Growth Metrics

Growth rates - both year-on-year (YoY) and quarter-on-quarter (QoQ) - are useful for trend analysis. The year-on-year (YoY) changes for various key series were computed and added to a `metrics_growth` table. 

In [22]:
run_sql_file(SQL_DIR / "derived" / "create table - metrics_growth.sql")
metrics_growth = run_sql("SELECT * FROM metrics_growth")
display(metrics_growth.tail(5))

,Ticker,ReportDate,RevenueGrowthYOY,GrossProfitGrowthYOY,OperatingIncomeGrowthYOY,NetIncomeGrowthYOY,TotalAssetsGrowthYOY,TotalLiabilitiesGrowthYOY,CashAndEquivGrowthYOY,NetCashFromOperatingActivitiesYOY,NetCashFromInvestingActivitiesYOY,NetCashFromFinancingActivitiesYOY
16892,ZYXI,2020-12-31,None,None,None,None,None,None,None,None,None,None
16893,ZYXI,2021-12-31,-0.8998,None,None,None,0.8465,2.8740,0.0878,7.4951,15.8589,-0.4824
16894,ZYXI,2022-12-31,18.6921,None,None,None,-0.1261,-0.1557,-0.5273,0.9781,-0.9748,-3.7334
16895,ZYXI,2023-12-31,0.1654,None,-0.5300,-0.4291,0.1755,0.8116,1.2130,0.2917,1.3110,-1.2136
16896,ZYXI,2024-12-31,0.0436,1.4388,-0.4443,-0.6924,-0.1083,-0.0485,-0.1110,-0.2836,-0.4017,-3.2356


## Profitability Ratios

Profitability ratios measure a company’s ability to convert revenue and assets into profit. They reveal the strength of the firm’s core business model and its capacity to generate returns for shareholders. Metrics like gross margin, operating margin, net margin, ROA, and ROE help distinguish companies that merely grow revenue from those that grow value. These ratios also highlight competitive advantages, pricing power, cost discipline, and long‑term sustainability. In downstream analysis, profitability ratios often serve as leading indicators of strategic success and capital allocation effectiveness.

The profitability ratios will include:
- Calculate:
- EffectiveTaxRate
- EnterpriseValue
- GrossMargin
- OperatingMargin
- ProfitMargin
- ROA
- ROE
- ROIC
- EBITToEV

In [23]:
run_sql_file(SQL_DIR / "derived" / "create table - metrics_profitability_annual.sql")
metrics_profitability_annual = run_sql("SELECT * FROM metrics_profitability_annual")
display(metrics_profitability_annual.tail(5))

,Ticker,ReportDate,EffectiveTaxRate,EnterpriseValue,GrossMargin,OperatingMargin,ProfitMargin,ROA,ROE,ROIC,EBITToEV
16892,ZYXI,2020-12-31,0.1932,None,0.7826,0.1279,0.1133,None,None,-39172999.85456677,None
16893,ZYXI,2021-12-31,None,None,0.8677,-0.5964,-0.8389,-0.0934,-0.1185,None,None
16894,ZYXI,2022-12-31,0.2329,None,0.7977,0.1450,0.1078,0.1279,0.2306,-9517999.73484959,None
16895,ZYXI,2023-12-31,0.3708,None,0.7919,0.0585,0.0528,0.0836,0.1467,13026000.14704109,None
16896,ZYXI,2024-12-31,None,None,0.7950,0.0311,0.0156,0.0219,0.0649,None,None


And quarterly ratios:

In [24]:
metrics_profitability_quarterly = run_sql_file(SQL_DIR / "derived" / "create table - metrics_profitability_quarterly.sql")
metrics_profitability_quarterly = run_sql("SELECT * FROM metrics_profitability_quarterly")
display(metrics_profitability_quarterly.tail(5))

,Ticker,ReportDate,EffectiveTaxRate,EnterpriseValue,GrossMargin,OperatingMargin,ProfitMargin,ROA,ROE,ROIC,EBITToEV
47640,ZYXI,2024-03-31,0.2308,None,0.8002,0.0113,0.0002,None,None,0.00695351,None
47641,ZYXI,2024-06-30,0.2464,None,0.8001,0.0446,0.0244,None,None,0.02795802,None
47642,ZYXI,2024-09-30,0.2502,None,0.7963,0.0761,0.0477,None,None,0.05048620,None
47643,ZYXI,2024-12-31,None,None,0.7829,-0.0122,-0.0134,-0.0085,-0.0108,None,None
47644,ZYXI,2025-03-31,None,None,0.6851,-0.4928,-0.3912,None,None,None,None


## Efficiency (Activity) Ratios

Efficiency ratios assess how effectively a company uses its assets and manages its operations. They highlight the speed and productivity with which resources are converted into revenue. Ratios such as asset turnover, inventory turnover, receivables turnover, and payables turnover illuminate operational discipline, supply‑chain management, and customer‑collection performance. These metrics often uncover bottlenecks or strengths that profitability ratios alone cannot reveal. Efficiency ratios are especially valuable for operational benchmarking and optimization.

In [25]:
run_sql_file(SQL_DIR / "derived" / "create table - metrics_activity.sql")
metrics_activity = run_sql("SELECT * FROM metrics_activity")
metrics_activity.tail(5)

,Ticker,ReportDate,SalesTurnover
47640,ZYXI,2024-03-31,0.3399
47641,ZYXI,2024-06-30,0.3952
47642,ZYXI,2024-09-30,0.4098
47643,ZYXI,2024-12-31,0.3647
47644,ZYXI,2025-03-31,0.2177


## Valuation Ratios

Valuation ratios connect fundamentals to market expectations. They measure how investors price a company relative to its earnings, cash flow, book value, or sales. Ratios like P/E, P/B, P/S, and EV/EBITDA help determine whether a stock appears undervalued, fairly valued, or overvalued. These metrics reflect sentiment, growth expectations, and perceived risk. In downstream analysis, valuation ratios bridge the gap between internal performance and external market behavior, enabling more informed investment decisions.

In [26]:
run_sql_file(SQL_DIR / "derived" / "create table - metrics_value.sql")
metrics_value = run_sql("SELECT * FROM metrics_value")
metrics_value.tail(5)

,Ticker,ReportDate,PriceDate,PE,PS,PB
47640,ZUO,2021-04-30,2021-04-29,None,24.47353758,11.46270881
47641,ZYME,2020-12-31,2020-12-30,None,154.33392434,5.90574109
47642,ZYNE,2020-12-31,2020-12-30,None,None,1.61032348
47643,ZYNE,2021-03-31,2021-03-30,None,None,1.94646189
47644,ZYNE,2022-12-31,2022-12-30,None,None,0.50154594


## Liquidity Ratios

Liquidity ratios evaluate a company’s ability to meet short‑term obligations using its most liquid assets. They answer a fundamental question: Can the company pay its bills without stress? Ratios such as the current ratio, quick ratio, and cash ratio help identify operational resilience and working‑capital efficiency. Strong liquidity supports stability during downturns, while weak liquidity can signal cash‑flow strain or reliance on external financing. These ratios are essential for risk assessment, especially in volatile industries or early‑stage companies.

In [27]:
run_sql_file(SQL_DIR / "derived" / "create table - metrics_liquidity.sql")
liquidity = run_sql("SELECT * FROM metrics_liquidity")
liquidity.tail(5)

,Ticker,ReportDate,CashRatio,QuickRatio,CurrentRatio
52415,ZYXI,2024-03-31,1.510503,2.679844,3.584877
52416,ZYXI,2024-06-30,1.694045,2.987718,4.063768
52417,ZYXI,2024-09-30,1.894000,2.982333,3.943527
52418,ZYXI,2024-12-31,2.350036,3.418703,4.457958
52419,ZYXI,2025-03-31,1.461788,2.323037,3.457376


## Solvency Ratios

Solvency ratios measure long‑term financial stability and the degree to which a company relies on debt to finance operations. Metrics like debt‑to‑equity, interest coverage, and debt‑to‑assets reveal leverage levels and the firm’s ability to service debt over time. High leverage can amplify returns but also magnify risk. Solvency ratios help analysts understand capital structure, creditworthiness, and exposure to macroeconomic conditions. They are critical for evaluating long‑term viability and strategic flexibility.

In [28]:
run_sql_file(SQL_DIR / "derived" / "create table - metrics_solvency.sql")
metrics_solvency = run_sql("SELECT * FROM metrics_solvency")
metrics_solvency.tail(5)

,Ticker,ReportDate,DebtToEquity,DebtRatio,InterestCoverageRatio
47644,ZUO,2021-04-30,0.4051,0.5859,None
47645,ZYME,2020-12-31,0.0211,0.2388,None
47646,ZYNE,2020-12-31,0.0000,0.1879,None
47647,ZYNE,2021-03-31,0.0000,0.1203,None
47648,ZYNE,2022-12-31,0.0000,0.1674,None


## Cash Flow Quality Ratios

Cash flow quality reflects how reliably a company turns its accounting earnings into actual cash, offering a deeper look into the sustainability of its operations beyond reported profit figures; metrics such as Earnings Quality, FCF Margin, and FCF Conversion reveal whether earnings are supported by cash generation or distorted by accruals, capital intensity, or working‑capital swings, and across the companies shown, ZUO demonstrates weak cash flow quality with negative earnings support and limited free‑cash generation, ZYME shows unusually strong earnings quality but poor FCF margin - often a sign of heavy investment or non‑recurring cash movements - and ZYNE consistently converts earnings into cash effectively despite missing FCF margin data, suggesting disciplined cash management; overall, cash flow quality acts as a critical lens for distinguishing durable performance from accounting noise, identifying firms capable of self‑funding growth, and assessing resilience when paired with profitability, efficiency, liquidity, and solvency ratios.

In [29]:
run_sql_file(SQL_DIR / "derived" / "create table - metrics_earnings_quality.sql")
metrics_earnings_quality = run_sql("SELECT * FROM metrics_earnings_quality")
metrics_earnings_quality.tail(5)

,Ticker,ReportDate,EarningsQuality,FCFMargin,FCFConversion
47642,ZUO,2021-04-30,-0.5804,0.1032,-0.4692
47643,ZYME,2020-12-31,1.2642,-3.1242,1.2912
47644,ZYNE,2020-12-31,0.7264,None,0.7329
47645,ZYNE,2021-03-31,1.0416,None,1.0416
47646,ZYNE,2022-12-31,0.7093,None,0.7102


# Conclusion

By converting raw fundamentals into a structured suite of ratios, this project creates a powerful analytical engine for understanding corporate performance. Each ratio group provides a different lens - profitability shows value creation, liquidity shows short‑term stability, solvency shows long‑term risk, efficiency shows operational execution, and valuation shows market perception. Together, they form a holistic framework that supports modeling, screening, forecasting, and strategic insights. This ratio‑driven foundation ensures that downstream analysis is not only data‑rich but also economically meaningful.

# Further Ideas
The data output from this project are informative and set the stage for further analysis. However, there are additional improvements that could be made to improve the efficiency of the data acquisition, processing, and output quality. They include:
1. Fetching the raw data from SimFin via its native API. This could involve automated download, unzipping, and folder cleanup.
2. Handling outlier values. For example, the stock prices contained extreme, probably incorrect data values. These could be deleted to ensure they do not affect subsequent calculations.
3. Handling financial report restatements. Occasionally, financial report restatements occur and the date of the restatement should become the effective date of the reported financial figures.